### ISSUE TITLE
Online lm_eval VEP metric reports chance for all models; migrate to AUPRC + RC averaging

### ISSUE THREAD

**Issue by gonzalobenegas** [Open-Athena/marin-dna#225]
## Summary

The **online (in-training) lm_eval** mendelian VEP metric reports **chance-level PairwiseAccuracy for every model**, even ones that are clearly skillful offline. It is not measuring the trained model. We want to **fix it by migrating the online task to AUPRC + cluster-bootstrap SE** (matching the offline `evals_v2` pipeline after #194/#195), **keeping the FWD+RC strand averaging**.

## Evidence (exp187, #187)

Online WandB `lm_eval/mendelian_traits_255/_global_/avg/pairwise_accuracy` across all 10 eval steps (500→4999), per arm:

| arm | online global PA (min–max) | offline AUPRC, same final ckpt |
|---|--:|--:|
| v3_cds | 0.476 – 0.503 | **0.311** missense / 0.233 global |
| v3_utr3 | 0.470 – 0.476 | — |
| v3_ncrna_exon | 0.470 – 0.508 | — |
| v3_tss_region_and_utr5 | 0.470 – 0.490 | — |
| v3_ccre_non_promoter | 0.471 – 0.502 | — |
| v3_bg | 0.471 – 0.486 | — |

**v3_cds is the tell:** online PA never leaves 0.48–0.50 across all of training, while the offline `evals_v2` run on the *same checkpoints* shows it learning monotonically (missense AUPRC 0.31, global 0.23, both many SE above the 0.10 1:9 baseline; PA-equivalent was ~0.76 on the pre-#194 1:1 dataset). Same model, same variants — the online path is broken. (macro_avg drifts to 0.51–0.58, which is just small-subset noise, not signal.)

## Why fix-by-AUPRC

Two things are true at once and we want both addressed in one change:

1. **The online metric is currently `pairwise_accuracy`** ([`mendelian_traits_255.yaml`](https://github.com/Open-Athena/marin-dna/blob/a8c28e1f351799895ace55697fad789b99f259c5/src/bolinas/pipelines/evals/lm_eval/mendelian_traits_255.yaml), [`dna_vep_llr_eval.py`](https://github.com/Open-Athena/marin-dna/blob/a8c28e1f351799895ace55697fad789b99f259c5/src/bolinas/pipelines/evals/lm_eval/dna_vep_llr_eval.py)), which #195 already retired offline because PA is invalid on the new 1:9 (k=9) matched-pair datasets. Online must move to **AUPRC** to be comparable to the leaderboard and the benchmarks.
2. The online task is *also* producing chance output (root cause TBD — see below), so simply swapping the metric is not enough; the fix must verify the scoring path actually works end-to-end.

**Keep FWD+RC averaging** — the online aggregator already averages the two strand-tagged rows per variant before the metric (`_PairwiseAccuracyAggregation`); the AUPRC replacement must preserve that (average raw LLR across FWD+RC, then transform, then AUPRC — mirroring `evals_v2` `metrics.smk` `minus_llr_avg`).

## Root cause — to be determined in the fix PR

Filing with evidence now; RCA during implementation. Likely suspects to check:
- **Dataset wiring** — the harness dataset is `bolinas-dna/evals_mendelian_traits_harness_255` (a *different* repo from the offline `bolinas-dna/evals_mendelian_traits`); verify its rows, the FWD/RC strand tags, and `match_group` are what `_PairwiseAccuracyAggregation` expects.
- **Scoring path** — `construct_requests` / `process_results` log-prob extraction: is `llr = log_prob_alt − log_prob_ref` reading the right token spans? A chance result smells like ref/alt log-probs being equal, swapped, or computed over the wrong positions.
- **Strand handling** — if FWD and RC scores are being mis-paired or one strand is constant, the per-variant average collapses to noise.
- **Context window** — 255 bp + BOS vs the harness dataset's encoding.

## Fix

Migrate the online lm_eval VEP task to **AUPRC + cluster-bootstrap SE** (port `compute_auprc_metrics` / `auprc_with_bootstrap_se` from `src/bolinas/pipelines/evals/metrics.py`, used offline since #195), retaining per-variant FWD+RC averaging and the `_global_` / `_macro_avg_` aggregation shape. Fix whatever scoring-path bug the RCA surfaces.

## Acceptance criterion — parity vs offline `evals_v2`

Online AUPRC on a known checkpoint must match the offline `evals_v2` AUPRC **within bootstrap SE, per subset**. Concrete target (`exp187-v3_cds-step-4999`, `minus_llr_avg`, from the 30 parquets already on S3):

| subset | offline AUPRC ± SE |
|---|--:|
| missense_variant | 0.311 ± 0.016 |
| _global_ | 0.233 ± 0.007 |
| _macro_avg_ | 0.225 ± 0.010 |

Run the online task on that exact checkpoint and confirm per-subset agreement. (The offline parquets at `s3://oa-bolinas/snakemake/analysis/evals_v2/results/metrics/exp187-*-step-*/mendelian_traits.parquet` are the ground truth.)

## Context

Surfaced in #187 (per-region sweep): the diagonal-wins analysis had to be done entirely from offline `evals_v2` because the online WandB metric was unusable. Fixing this restores cheap in-training tracking of VEP skill.

**Comment by gonzalobenegas** [comment]
🤖 **Clarifications before implementing**

After reading the code and inspecting the live datasets, three corrections to the framing above. None change the conclusion (migrate the online metric to AUPRC), but they retarget the work and sharpen the acceptance check.

**1. The evidence table compares two different metrics.** It pairs *offline AUPRC* against *online PairwiseAccuracy*. Those aren't comparable, so the flat ~0.5 online PA is not by itself evidence of a broken scorer — it's a different metric on a different scale (PA chance = 0.5; AUPRC chance = prevalence ≈ 0.1 on a 1:9 dataset). The apples-to-apples check is online-AUPRC vs offline-AUPRC (what this issue delivers), or online-PA vs offline-PA.

**2. The online path has never used AUPRC.** `git log` on `dna_vep_llr_eval.py` shows only #179 (which *added* PairwiseAccuracy + FWD/RC averaging) and the rename — there's no prior AUPRC code in the online task to "migrate back to." The actual history of the online metric:

| version | metric | FWD/RC averaging |
|---|---|---|
| v0 | per-row AUPRC | no |
| v1 (#179, current) | PairwiseAccuracy | yes |
| **target (this issue)** | **AUPRC** | **yes** |

So the task is to **combine the two pieces that each already worked** — v0's AUPRC scoring and v1's per-variant FWD/RC averaging — not to root-cause a novel scoring bug. That also lowers the risk that there's a deep scoring-path bug hiding here.

**3. Both datasets are now 1:9, and the current code can no longer run on the harness dataset.** When the matched datasets moved to 1:9 (k=9, #194), the harness dataset (`evals_mendelian_traits_harness_255`) was re-materialized too (HF commit `7b92f04`, 2026-05-19). The current v1 aggregator calls `pairwise_accuracy`, which asserts exactly 1 positive + 1 negative per `match_group`; on the 1:9 frame each group has 1 positive + 9 negatives, so it would now **assert-crash**. The 0.48–0.50 PA in the table predates that re-upload (computed on the old 1:1 dataset). The offline `mendelian_traits` and the harness dataset now share the **same 1614 match_groups / 9 subsets / variants**, so once the online metric is AUPRC the two are directly comparable.

## Plan

- Migrate the online task to **FWD/RC-avg + AUPRC**: average **raw LLR** across the two strand rows per variant → apply the `minus_llr` transform → `compute_auprc_metrics` (the same offline helper; `n_bootstrap=1000`, `seed=0`, `n_min=30`, so the SE matches offline for free). Headline scalar becomes `_global_/avg/auprc`.
- **Pin the HF dataset commit** in the online task (`revision=7b92f04…`) so the dataset can't silently change under the eval again — which is exactly the failure mode here (1:1 → 1:9 re-upload).

## Validation

- **Train a fresh model for 10 steps with the online lm_eval enabled** — validate in the *real in-training setting* (the path that was unusable in #187), not an eval-only harness on a pre-trained checkpoint. Then run offline `evals_v2` on that exact 10-step HF checkpoint and confirm **online AUPRC == offline AUPRC per subset, within bootstrap SE**.
- At 10 steps the model is ~untrained, so both sides sit near the AUPRC baseline (≈0.1 `_global_`); the test is *agreement between the two paths on the same checkpoint*, which a scoring bug would break regardless of skill.
- Reference for a *converged* model (sanity anchor, not the parity target): `exp187-v3_cds-step-4999` offline — missense 0.311 ± 0.016, `_global_` 0.233 ± 0.007, `_macro_avg_` 0.225 ± 0.010.

<details><summary>Verification details (datasets, git, bootstrap config)</summary>

**Live HF datasets (train split), both last modified 2026-05-19:**

```
evals_mendelian_traits_harness_255 (online):  32280 rows; target True/False = 3228/29052
  strand: {'+': 16140, '-': 16140}   → 16140 variants × 2 strands
  per-match_group: 20 rows each, 2 positives each (1 variant × 2 strands)
  1614 match_groups, 9 subsets   HEAD commit 7b92f04

evals_mendelian_traits (offline):  16140 rows; label True/False = 1614/14526
  per-match_group: 10 rows each, 1 positive each   → 1:9
  1614 match_groups, 9 subsets   HEAD commit 4aed58e
```

**git** — `dna_vep_llr_eval.py` history: `a8c28e1` rename only; PA introduced in #179 per `task_configs.py` ("PairwiseAccuracy + per-variant FWD/RC averaging replaces the old per-row AUPRC"). No AUPRC ever in `src/.../lm_eval/`.

**Offline bootstrap config** (`snakemake/analysis/evals_v2/config/config.yaml`): `n_bootstrap: 1000`, `bootstrap_seed: 0`, `split: train` — identical to `compute_auprc_metrics` defaults.

**exp187-v3_cds-step-4999 offline parquet** (`minus_llr_avg`): missense 0.311396 ± 0.015858 (n_groups 580), _global_ 0.233322 ± 0.007299 (1614), _macro_avg_ 0.225243 ± 0.010424 (8 qualifying subsets). Checkpoint: `gs://marin-us-east5/checkpoints/dna-exp187-zoonomia-v1-1b-v3_cds-v0.1-d6940d/hf/step-4999` (Qwen3-1B, HF format).

</details>

**Comment by gonzalobenegas** [comment]
🤖 **Validation result — online AUPRC == offline AUPRC (parity confirmed)**

Implemented in #226 and validated in the **real in-training setting** (per the plan above): a tiny 10-step Qwen3 run (`experiments/parity/issue225_online_auprc_parity.py`) with the `mendelian_traits_255` lm-eval task firing in-training, compared head-to-head against offline `evals_v2` run on the **same HF checkpoints**. Both sides use `compute_auprc_metrics` (`n_bootstrap=1000`, `seed=0`, `n_min=30`), so the SEs are directly comparable.

**Result: online == offline within bootstrap SE, across every subset, at both step-5 and step-9.** Nothing exceeds `|Δ| > SE`.

Headline subsets (step-9, the final checkpoint):

| subset | offline `minus_llr_avg` | online `_/avg/auprc` | Δ |
|---|--:|--:|--:|
| _global_ | 0.0941 ± 0.0023 | 0.0937 ± 0.0022 | +0.0004 |
| missense_variant | 0.0975 ± 0.0044 | 0.0966 ± 0.0040 | +0.0010 |
| _macro_avg_ | 0.1056 ± 0.0053 | 0.1032 ± 0.0047 | +0.0024 |

<details><summary>Full per-subset tables (step-5 and step-9)</summary>

**step-5** (`…/hf/step-5`)

| subset | offline | online | Δ |
|---|--:|--:|--:|
| _global_ | 0.1000 ± 0.0026 | 0.0996 ± 0.0026 | +0.0003 |
| _macro_avg_ | 0.1124 ± 0.0068 | 0.1089 ± 0.0056 | +0.0035 |
| missense_variant | 0.1086 ± 0.0060 | 0.1081 ± 0.0060 | +0.0005 |
| splicing | 0.0891 ± 0.0035 | 0.0902 ± 0.0039 | −0.0011 |
| 5_prime_UTR_variant | 0.1021 ± 0.0079 | 0.0999 ± 0.0078 | +0.0022 |
| 3_prime_UTR_variant | 0.0895 ± 0.0076 | 0.0888 ± 0.0073 | +0.0008 |
| non_coding_transcript_exon_variant | 0.1033 ± 0.0117 | 0.1166 ± 0.0168 | −0.0133 |
| tss_proximal | 0.0972 ± 0.0063 | 0.0952 ± 0.0058 | +0.0020 |
| distal | 0.1625 ± 0.0333 | 0.1383 ± 0.0266 | +0.0242 |
| synonymous_variant | 0.1471 ± 0.0389 | 0.1344 ± 0.0292 | +0.0127 |
| mature_miRNA_variant | 0.1199 ± 0.0316 | — (n_groups=4 < 30, not logged) | — |

**step-9** (`…/hf/step-9`, final)

| subset | offline | online | Δ |
|---|--:|--:|--:|
| _global_ | 0.0941 ± 0.0023 | 0.0937 ± 0.0022 | +0.0004 |
| _macro_avg_ | 0.1056 ± 0.0053 | 0.1032 ± 0.0047 | +0.0024 |
| missense_variant | 0.0975 ± 0.0044 | 0.0966 ± 0.0040 | +0.0010 |
| splicing | 0.0814 ± 0.0029 | 0.0817 ± 0.0031 | −0.0003 |
| 5_prime_UTR_variant | 0.1000 ± 0.0083 | 0.1027 ± 0.0099 | −0.0027 |
| 3_prime_UTR_variant | 0.1064 ± 0.0167 | 0.1071 ± 0.0188 | −0.0008 |
| non_coding_transcript_exon_variant | 0.1254 ± 0.0161 | 0.1260 ± 0.0158 | −0.0006 |
| tss_proximal | 0.0898 ± 0.0050 | 0.0918 ± 0.0065 | −0.0020 |
| distal | 0.0883 ± 0.0085 | 0.0861 ± 0.0084 | +0.0022 |
| synonymous_variant | 0.1563 ± 0.0327 | 0.1337 ± 0.0240 | +0.0226 |
| mature_miRNA_variant | 0.1188 ± 0.0640 | — (n_groups=4 < 30, not logged) | — |

</details>

**Notes**
- The in-training path ran on the 1:9 harness **without crashing** — the old PA aggregator asserts exactly 1 positive + 1 negative per `match_group`, so it would have crashed on the 1:9 dataset. It now emits per-subset + `_global_` + `_macro_avg_` × fwd/rc/avg AUPRC + SE cells to WandB (headline `_global_/avg/auprc`).
- The HF dataset commit is now **pinned** (`revision=7b92f04…`), closing the silent 1:1 → 1:9 drift that produced the misleading "online PA ≈ chance" evidence originally filed here.
- **Caveat:** at 10 steps the model is ~untrained, so both sides sit at the ≈0.1 AUPRC baseline. This validates wiring / metric / `n_min` / strand-averaging agreement + no-crash on 1:9; sign/span scoring correctness rides on the **unchanged** (and unit-tested) LLR path — not re-proven on a skillful model here.
- WandB run: `dna-exp225-online-auprc-parity-tiny-v0.1` (project `marin`).




---

**Update — point-only online (`f119af4`).** Per the PR self-review, the in-training metric now computes **point AUPRC only**: the cluster-bootstrap SE is dropped online (redundant with this offline parquet, and a ~tens-of-seconds/eval-step cost on the train loop) and stays the authoritative *offline* number. The point values in the tables above are **unchanged** — the bootstrap only affects the SE, not the AUPRC — so the parity still holds.
